Add except_ref type

aheejin opened the pull request:
This adds [except_ref](https://github.com/WebAssembly/exception-handling/blob/master/proposals/Exceptions.md#except_ref) type, which is a part of the [exception handling proposal](https://github.com/WebAssembly/exception-handling/blob/master/proposals/Exceptions.md).

kripken commented on src/passes/ConstHoisting.cpp:
```diff
@@ -95,6 +95,8 @@ struct ConstHoisting : public WalkerPass<PostWalker<ConstHoisting>> {
         // v128 not implemented yet
         return false;
       }
+      case except_ref: // except_ref cannot have literal
```
we usually have `{ }` around case bodies, but sadly not consistently. looks like we do on tihs particular switch though, so may be nice to keep it at least consistent.

kripken commented on src/passes/InstrumentLocals.cpp:
```diff
@@ -57,11 +57,13 @@ Name get_i32("get_i32");
 Name get_i64("get_i64");
```
can an `except_ref` be passed to JS? (how does it look there?)

kripken commented on src/tools/fuzzing.h:
```diff
@@ -1676,6 +1685,10 @@ class TranslateToFuzzReader {
   Expression* makeConst(Type type) {
+    // There's no except_ref.const. Returns something else instead.
+    if (type == except_ref) {
+      return makeGetLocal(type);
```
I think this can infinitely recurse :( if we have no local of that type, it will call this `makeConst` here...

Instead, one option is to add a new local of that type in makeGetLocal. Another is to have a singleton global which we use for this purpose.

aheejin replied:
In the [reference type proposal](https://github.com/WebAssembly/reference-types/blob/master/proposals/reference-types/Overview.md), there's an instruction named `ref.null` which returns a nullref value. We can use that here after we implement those instructions, but can we leave it as TODO for now?

aheejin commented:
> * I'm not sure what capitalization we should use for `except_ref`. Perhaps `exceptRef`? Or maybe `ExceptRef`?

How about `ExceptRef`? I noticed that we have `CamelCase` style naming in [`BinaryConsts::EncodedType`](https://github.com/WebAssembly/binaryen/blob/415e8b37c2e570117b17394d4910ca57fa2d11d8/src/wasm-binary.h#L373-L386), such as `AnyFunc`. We need to add the exception ref type here as well, and adding `ExceptRef` here and also to the wasm type in wasm.h looks more consistent. Changed to `ExceptRef`, but please let me know if you still prefer `exceptRef`.

> * We have `bool isConcreteType(Type type); bool isFloatType(Type type); bool isIntegerType(Type type); bool isVectorType(Type type);` in `wasm-type.h`. I wonder if maybe we should add `is???Type` for a "GC type", which might include except_ref and future any_ref and other things? Then if we use `is?GC?Type()` we won't need to update all the places again for future "GC" types.

Added `isReferenceType(Type type)`.

> * Please add a test file `test/except_ref.wast` that contains the basics here. Then the test suite will add the output of converting that to binary and back, for test coverage of the text and binary parsing code.

Added in `test/exception-handling.wast`, to which I plan to add more EH related tests. Also added the parsing routines to the s parser and the binary parser for the new type.

aheejin commented:
Hmm, come to think of it, because we already have `none` and `unreachable` in wasm types, would it be better to change it back to `exceptRef`? I'm not sure...

tlively commented:
I would lean toward keeping it `except_ref` for now actually for consistency with the other types (C++ name exactly matching wast name) and then rename them all in a separate PR if we want to change their style.

kripken commented on src/tools/fuzzing.h:
```diff
@@ -815,6 +815,7 @@ class TranslateToFuzzReader {
       case f32:
```
I think the fuzzing support is just not present yet, in that `getType` doesn't return an except_ref? (that is, no except_refs will be generated by the fuzzer) probably worth adding a clear TODO on that near the top somewhere (as otherwise, the preparations for except_ref support scattered throughout might imply that it does work)

aheejin replied:
Done
